replications/remotewrite/writer.go

```diff
@@ -0,0 +1,242 @@
+		Err:  err,
+	}
+}
+
+func invalidResponseCode(code int) *ierrors.Error {
+	return &ierrors.Error{
+		Code: ierrors.EInvalid,
+		Msg:  fmt.Sprintf("invalid response code %d, must be %d", code, http.StatusNoContent),
+	}
+}
+
+type HttpConfigStore interface {
+	GetFullHTTPConfig(context.Context, platform.ID) (*influxdb.ReplicationHTTPConfig, error)
+	UpdateResponseInfo(context.Context, platform.ID, int, string) error
+}
+
+type waitFunc func(time.Duration) <-chan time.Time
+
+type writer struct {
+	replicationID                 platform.ID
+	configStore                   HttpConfigStore
+	metrics                       *metrics.ReplicationsMetrics
+	logger                        *zap.Logger
+	maximumBackoffTime            time.Duration
+	maximumAttemptsForBackoffTime int
+	clientTimeout                 time.Duration
+	maximumAttemptsBeforeErr      int      // used for testing, 0 for unlimited
+	waitFunc                      waitFunc // used for testing
+}
+
+func NewWriter(replicationID platform.ID, store HttpConfigStore, metrics *metrics.ReplicationsMetrics, logger *zap.Logger) *writer {
+	return &writer{
+		replicationID:                 replicationID,
+		configStore:                   store,
+		metrics:                       metrics,
+		logger:                        logger,
+		maximumBackoffTime:            maximumBackoffTime,
+		maximumAttemptsForBackoffTime: maximumAttempts,
+		clientTimeout:                 DefaultTimeout,
+		waitFunc: func(t time.Duration) <-chan time.Time {
+			return time.After(t)
+		},
+	}
+}
+
+func (w *writer) Write(ctx context.Context, data []byte) error {
+	attempts := 0
+
+	for {
+		if w.maximumAttemptsBeforeErr > 0 && attempts >= w.maximumAttemptsBeforeErr {
+			return errors.New("maximum number of attempts exceeded")
+		}
+
+		// Get the most recent config on every attempt, in case the user has updated the config to correct errors.
+		conf, err := w.configStore.GetFullHTTPConfig(ctx, w.replicationID)
+		if err != nil {
+			return err
+		}
+
+		res, err := PostWrite(ctx, conf, data, w.clientTimeout)
+		res, msg, ok := normalizeResponse(res, err)
+		if !ok {
+			// Can't retry - bail out.
+			return err
+		}
+
+		if err := w.configStore.UpdateResponseInfo(ctx, w.replicationID, res.StatusCode, msg); err != nil {
+			return err
+		}
+
+		if err == nil {
+			// Successful write
+			return nil
+		}
+
+		attempts++
+		var waitTime time.Duration
+		hasSetWaitTime := false
+
+		switch res.StatusCode {
+		case http.StatusBadRequest:
+			if conf.DropNonRetryableData {
+				w.logger.Debug(fmt.Sprintf("dropped %d bytes of data due to %d response from server", len(data), http.StatusBadRequest))
+				return nil
+			}
+		case http.StatusTooManyRequests:
+			headerTime := w.waitTimeFromHeader(res)
+			if headerTime != 0 {
+				waitTime = headerTime
+				hasSetWaitTime = true
+			}
+		}
+
+		if !hasSetWaitTime {
+			waitTime = w.backoff(attempts)
+		}
+
+		select {
+		case <-w.waitFunc(waitTime):
```

Comment:
This select statement isn't doing a whole lot now, but I think it will be useful for follow-up issues like https://github.com/influxdata/influxdb/issues/22943